### ISSUE TITLE
Pulling metrics, but no data output

### ISSUE THREAD

**Issue by dennis-bsi** [logstash-plugins/logstash-input-cloudwatch#2]
When I try to use the plugin, although it seems to retrieve the metrics, they never get outputted by logstash.  I added a few extra logger.info lines to try to troubleshoot, but I just don't know enough about the plugin architecture to get far.

```
Polling CloudWatch API {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"133", :method=>"run"}
[AWS CloudWatch 200 2.694 0 retries] list_metrics(:namespace=>"AWS/EC2")  
 {:level=>:info, :file=>"aws/core/client.rb", :line=>"410", :method=>"log_response"}
Polling metric CPUUtilization {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"139", :method=>"run"}
Filters: [{:name=>"tag:Monitoring", :values=>["Yes"]}] {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"140", :method=>"run"}
Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"450", :method=>"flush"}
Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"450", :method=>"flush"}
[AWS EC2 200 1.079 0 retries] describe_instances(:filters=>[{:name=>"tag:Monitoring",:values=>["Yes"]}])  
 {:level=>:info, :file=>"aws/core/client.rb", :line=>"410", :method=>"log_response"}
AWS/EC2 Instances: ["i-7ec3f7dd"] {:level=>:debug, :file=>"logstash/inputs/cloudwatch.rb", :line=>"264", :method=>"resources"}
Polling resource InstanceId: i-7ec3f7dd {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"180", :method=>"fetch_resource_events"}
[AWS CloudWatch 200 0.678 0 retries] get_metric_statistics(:dimensions=>[{:name=>"InstanceId",:value=>"i-7ec3f7dd"}],:end_time=>"2016-02-17T18:49:41Z",:metric_name=>"CPUUtilization",:namespace=>"AWS/EC2",:period=>300,:start_time=>"2016-02-17T18:44:41Z",:statistics=>["SampleCount","Average","Minimum","Maximum","Sum"])  
 {:level=>:info, :file=>"aws/core/client.rb", :line=>"410", :method=>"log_response"}
DPs: {:datapoints=>[{:timestamp=>2016-02-17 18:44:00 UTC, :sample_count=>5.0, :unit=>"Percent", :minimum=>1.02, :maximum=>4.25, :sum=>10.66, :average=>2.132}], :label=>"CPUUtilization", :response_metadata=>{:request_id=>"343c9b68-d5a7-11e5-9777-4f5c72bb6080"}} {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"183", :method=>"fetch_resource_events"}
DPArray: 1 {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"184", :method=>"fetch_resource_events"}
Event: {:timestamp=>2016-02-17 18:44:00 UTC, :sample_count=>5.0, :unit=>"Percent", :minimum=>1.02, :maximum=>4.25, :sum=>10.66, :average=>2.132} {:level=>:info, :file=>"logstash/inputs/cloudwatch.rb", :line=>"186", :method=>"fetch_resource_events"}
Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"450", :method=>"flush"}
Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"450", :method=>"flush"}
Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"450", :method=>"flush"}
Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"450", :method=>"flush"}
```

**Comment by subhash12** [comment]
Hi

If I have stdout as output, Im able to see the cloudwatch stats on the console. But if the output is elasticsearch, then I'm getting the following error: 

Logstash startup completed
NoMethodError: private method `sprintf' called for #<Hash:0x6821552e>
  event_action_params at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.1.4-java/lib/logstash/outputs/elasticsearch/common.rb:112
   event_action_tuple at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.1.4-java/lib/logstash/outputs/elasticsearch/common.rb:32
              receive at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-output-elasticsearch-2.1.4-java/lib/logstash/outputs/elasticsearch/common.rb:22
               handle at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/outputs/base.rb:81
          output_func at (eval):26
         outputworker at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/pipeline.rb:276
        start_outputs at /opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-2.1.0-java/lib/logstash/pipeline.rb:193

Logstash: logstash-2.1.0
Elasticsearch: elasticsearch-2.1.0

I find already there is a compatibility issue raised for this plugin:
elastic/logstash#4375

Thanks
Subhash

**Comment by jrgns** [comment]
I tried replicating your issue (using Logstash 2.2 and Elasticsearch 2.2). I couldn't get the NoMethodError, but I did pick up something (fixed in #3) were certain metrics weren't pushed to the queue properly.

Can you try again with the latest version of the plugin?

**Comment by dennis-bsi** [comment]
I resolved my issues by adding some code to the fetch_resource_events function.  If you look, it creates the events, but never returns them anywhere.  I added an array events to the function and pushed the events into it.  I then returned this array and from_resources started doing its thing.

I will clean everything up and create a pull request for my changes.

**Comment by jrgns** [comment]
Yeah. The code is a bit confusing. `fetch_resource_events` is supposed to return an enumerator of events to `from_resources`, where it should then be added to the queue. That doesn't work as expected, though.

I've combined the two functions in PR #3 which solves the problem.

@suyograo can we look at gett #3 merged?

**Comment by subhash12** [comment]
@jrgns Im getting the results in elasticsearch after the changes from #3
Im seeing another issue. I had default configuration. ie, Interval of 15 mins and Period of 5 min.
the 'timestamp' field of the datasets has the same value as 'end-time' of the datasets. Shouldnt it be having values 5 mins apart..

**Comment by jrgns** [comment]
The `@timestamp` is set to the `end_time` of each metric that comes in, yes. The argument is that the end time is when the statistics were measured.

They should have values 5 minutes apart, each time they come in. What you might be seeing is a set of events, all with the same end time, but for different metrics (CPUUtilization, NetworkIn, etc.)

It might be helpful if you post some of the events here?

**Comment by subhash12** [comment]
the data from ec2 instance.

timestamp:2016-02-19 07:18:34 +0000 sample_count:5 unit:Percent minimum:0.83 maximum:0.96 sum:4.42 average:0.884 namespace:AWS/EC2 metric_name:CPUUtilization start_time:2016-02-19 07:03:34 +0000 end_time:2016-02-19 07:18:34 +0000 period:300 InstanceId:i-03005264 @version:1 @timestamp:February 19th 2016, 12:48:35.015 _id:AVL4ZQLEa3pOCuOxN8fY _type:logs _index:stats-2016.02.19 _score:
timestamp:2016-02-19 07:18:34 +0000 sample_count:5 unit:Percent minimum:0.82 maximum:1.29 sum:4.96 average:0.992 namespace:AWS/EC2 metric_name:CPUUtilization start_time:2016-02-19 07:03:34 +0000 end_time:2016-02-19 07:18:34 +0000 period:300 InstanceId:i-03005264 @version:1 @timestamp:February 19th 2016, 12:48:34.990 _id:AVL4ZQLEa3pOCuOxN8fX _type:logs _index:stats-2016.02.19 _score:

Basically, I must be getting 3 data-sets. Getting only 2.
I agree with @timestamp value. The first field 'timestamp' must be having values 5 minutes apart. But they are looking the same as end_time field.

**Comment by subhash12** [comment]
'timestamp' with below values, was expected 
07:08:34
07:13:34
07:18:34

**Comment by dennis-bsi** [comment]
I noticed the @timestamp issue also. The easiest fix is to make timestamp, which is the aws metric timestamp, go into elasticsearch as a datetime instead of a string.

**Comment by dennis-bsi** [comment]
@subhash12 I worked around the timestamp issue by adding `event[:timestamp]   = event[:timestamp].iso8601` into the cleanup(event) function.  Now in Kibana, timestamp (without '@') reflects what the metric time in Cloudwatch is.

**Comment by subhash12** [comment]
@dennis-bsi If we stringify 'timestamp', the datatype in elasticsearch would be String. If I want it as Date, should I exclude that field from stringify ?

**Comment by ark-elk** [comment]
@dennis-bsi I am running a twitter input example and on the log i see below issue. This line just keeps repeating. Also saw message about twitter/codec plugin not defined in namespace. could you pls help


Pushing flush onto pipeline {:level=>:debug, :file=>"logstash/pipeline.rb", :line=>"458", :method=>"flush"}

**Comment by ark-elk** [comment]
guys any comments ?

**Comment by NehaBansal1234** [comment]
I am using log stash output new relic plugin and when it tries to flush the events to New Relic, it gives  Failed to Flush Outgoing Items,  NET::HTTPServerException . Is anybody aware of this issue?
